helix-core/src/shellwords.rs

```diff
@@ -1,198 +1,457 @@
+    }
+
     /// Checks that the input ends with a whitespace character which is not escaped.
     ///
     /// # Examples
     ///
     /// ```rust
-    /// use helix_core::shellwords::Shellwords;
+    /// # use helix_core::shellwords::Shellwords;
     /// assert_eq!(Shellwords::from(" ").ends_with_whitespace(), true);
     /// assert_eq!(Shellwords::from(":open ").ends_with_whitespace(), true);
     /// assert_eq!(Shellwords::from(":open foo.txt ").ends_with_whitespace(), true);
     /// assert_eq!(Shellwords::from(":open").ends_with_whitespace(), false);
-    /// #[cfg(unix)]
-    /// assert_eq!(Shellwords::from(":open a\\ ").ends_with_whitespace(), false);
-    /// #[cfg(unix)]
+    /// assert_eq!(Shellwords::from(":open a\\ ").ends_with_whitespace(), true);
     /// assert_eq!(Shellwords::from(":open a\\ b.txt").ends_with_whitespace(), false);
     /// ```
+    #[inline]
     pub fn ends_with_whitespace(&self) -> bool {
-        matches!(self.state, State::OnWhitespace)
+        self.input.ends_with(' ')
     }
+}
+
+/// An iterator over an input string which yields arguments.
+///
+/// Splits on whitespace, but respects quoted substrings (using double quotes, single quotes, or backticks).
+#[derive(Debug, Clone, Copy)]
+pub struct Args<'a> {
+    input: &'a str,
+    idx: usize,
+    start: usize,
+}
 
-    /// Returns the list of shellwords calculated from the input string.
-    pub fn words(&self) -> &[Cow<'a, str>] {
-        &self.words
+impl<'a> Args<'a> {
+    #[inline]
+    fn parse(input: &'a str) -> Self {
+        Self {
+            input,
+            idx: 0,
+            start: 0,
+        }
     }
 
-    /// Returns a list of strings which correspond to [`Self::words`] but represent the original
-    /// text in the input string - including escape characters - without separating whitespace.
-    pub fn parts(&self) -> &[&'a str] {
-        &self.parts
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.input.is_empty()
+    }
+
+    /// Returns the args exactly as input.
+    ///
+    /// # Examples
+    /// ```
+    /// # use helix_core::shellwords::Args;
+    /// let args = Args::from(r#"sed -n "s/test t/not /p""#);
+    /// assert_eq!(r#"sed -n "s/test t/not /p""#, args.raw());
+    ///
+    /// let args = Args::from(r#"cat "file name with space.txt""#);
+    /// assert_eq!(r#"cat "file name with space.txt""#, args.raw());
+    /// ```
+    #[inline]
+    pub fn raw(&self) -> &str {
+        self.input
+    }
+
+    /// Returns the remainder of the args exactly as input.
+    ///
+    /// # Examples
+    /// ```
+    /// # use helix_core::shellwords::Args;
+    /// let mut args = Args::from(r#"sed -n "s/test t/not /p""#);
+    /// assert_eq!("sed", args.next().unwrap());
+    /// assert_eq!(r#"-n "s/test t/not /p""#, args.rest());
+    /// ```
+    ///
+    /// Never calling `next` and using `rest` is functionally equivalent to calling `raw`.
+    #[inline]
+    pub fn rest(&self) -> &str {
```

Comment:
I think this is genreally the right API but its not used for `:set` and `:toggle` yet. Particularly for `:toggle` this would solve a lot of issues since we can pass the `rest` directly to `serde_json` which will give you a stream of json value (serde_json understands json syntax and knows where a new value starts/ends): https://docs.rs/serde_json/latest/serde_json/struct.StreamDeserializer.html

Reply:
Something I have have been curious about is the error message for toggle when its a string:
`"Bad arguments. For number configurations use: `:toggle key val1 val2 ...`"`

The `...` implies you can provide more than two, as well as the check itself `args.count() >= 2`. What config option would this work for? From the docs, it appears to me that any multivalve config takes a list(which as a json Value isnt supported yet), like `shell`(which hopefully after this change is supported now).

Reply:
For toggle, I am initially implementing this only for Array, as the string input requires quotes in the raw text for it to parse a `Value::String` correctly. This is documented in a comment in `set` as well. `Value::Bool` doesn't use any actual value, it just uses the currently stored value, so no need to change anything here. I can implement this for `Value::Number`, but it doesn't really change anything, as the parsing is the same, but I can do so if you wish.

Reply:
yes toggle allows an arbitrary number of arguments and will essentially cicle between them but that requires we can deterimne wher a json value starts/ends

Also deuciding how to parse the values based on the current value because some options can switch between list and single variants. That's why I would love to use this method + streaming deserialization for that

Reply:
Hmm, I'm not really sure how to handle this correctly. If there are duplicate inputs the current `skip_while` would get stuck I think. This would affect both strings and numbers. And if there are an arbitrary number of arrays accepted, I'm only handling the first two arrays currently.